dpctl/tensor/libtensor/include/kernels/elementwise_functions/remainder.hpp

```diff
@@ -0,0 +1,336 @@
+    using supports_sg_loadstore = std::negation<
+        std::disjunction<tu_ns::is_complex<argT1>, tu_ns::is_complex<argT2>>>;
+    using supports_vec = std::conjunction<
+        std::is_same<argT1, argT2>,
+        std::negation<std::disjunction<tu_ns::is_complex<argT1>,
+                                       tu_ns::is_complex<argT2>>>>;
+
+    resT operator()(const argT1 &in1, const argT2 &in2)
+    {
+        if constexpr (std::is_integral_v<argT1> || std::is_integral_v<argT2>) {
+            if (in2 == 0) {
+                return 0;
+            }
+            if constexpr (std::is_unsigned_v<argT1> ||
+                          std::is_unsigned_v<argT2>) {
+                return (in1 % in2);
+            }
+            else {
+                auto out = (in1 % in2);
+                if (out != 0 && ((in1 < 0) != (in2 < 0))) {
+                    out += in2;
+                }
+                return out;
+            }
+        }
+        else {
+            auto rem = sycl::fmod(in1, in2);
+            if (rem) {
+                if ((in2 < 0) != (rem < 0)) {
+                    rem += in2;
+                }
+            }
+            else {
+                rem = std::copysign(resT(0), in2);
+            }
+            return rem;
+        }
+    }
+
+    template <int vec_sz>
+    sycl::vec<resT, vec_sz> operator()(const sycl::vec<argT1, vec_sz> &in1,
+                                       const sycl::vec<argT2, vec_sz> &in2)
+    {
+        if constexpr (std::is_integral_v<argT1> || std::is_integral_v<argT2>) {
+            auto rem = (in1 % in2);
+            if constexpr (std::is_unsigned_v<argT1> ||
+                          std::is_unsigned_v<argT2>) {
+                if constexpr (std::is_same_v<resT, typename decltype(
+                                                       rem)::element_type>) {
+                    return rem;
+                }
+                else {
+                    using dpctl::tensor::type_utils::vec_cast;
+
+                    return vec_cast<resT, typename decltype(rem)::element_type,
+                                    vec_sz>(rem);
+                }
+            }
+            else {
+#pragma unroll
+                for (auto i = 0; i < vec_sz; ++i) {
+                    if (in2[i] == 0) {
+                        rem[i] = 0;
+                    }
+                    if (rem[i] != 0 && ((in1[i] < 0) != (in2[i] < 0))) {
+                        rem[i] += in2[i];
+                    }
+                }
+                if constexpr (std::is_same_v<resT, typename decltype(
+                                                       rem)::element_type>) {
+                    return rem;
+                }
+                else {
+                    using dpctl::tensor::type_utils::vec_cast;
+
+                    return vec_cast<resT, typename decltype(rem)::element_type,
+                                    vec_sz>(rem);
+                }
```

Comment:
The PR adding bitwise functions sneaks in an optimization to `vec_cast` which just returns the input if `dstT` is same as `srcT`. With that change, we can just  always call `vec_cast` to make the code more readable. 

This change is best delayed to a separate PR once bitwise functions PR is reviewed and merged.